File: python/leetcode113.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if not root:
            return []
        if root.left == None and root.right == None:
            if root.val == sum:
                return [[root.val]]
            else:
                return []
        a = self.pathSum(root.left, sum-root.val) + self.pathSum(root.right, sum - root.val)
        return [[root.val]+i for i in a]

File: leetcodec++/leetcode113.cpp
//
// Created by 黄敏辉 on 2019-08-02.
//

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> res;
        vector<int> path;
        helper(root, res, path, sum);
        return res;
    }

    void helper(TreeNode* root, vector<vector<int>>& res, vector<int>& path, int sum){
        if (root == NULL) return;
        path.push_back(root->val);
        if (!root->left && !root->right && sum == root->val){
            res.push_back(path);
        }
        helper(root->left, res, path, sum - root->val);
        helper(root->right, res, path, sum - root->val);
        path.pop_back();
    }
};